### PR TITLE
Retain qnn input kernel scales

### PR DESCRIPTION
**Pull request by u99127**
The QNN dialect loses the input tensor scale and the weight
tensor scale too early. This inhibits the work one could do
with integrating 3rd party codegen or libraries that have interfaces
that expect input tensor scale and weight tensor scales.

This patch stack fixes it up for conv2d and Dense. I cannot see
any other operators affected yet. 

See here for more . https://discuss.tvm.ai/t/lowering-qnn-conv2d-tflite/4654/5

Ramana

@anijain2305 - Please review.

### CODE REVIEW

**Comment by u99127** [comment]
> 
> 
> Overall LGTM. The only change I request is changing the name to input_scale and kernel_scale. I have similar naming elsewhere, so will be easy to read.

Sure.

When I wrote it, I felt it better to keep the name distinct to indicate the difference between input_scale in Requantize vs here but I suppose `tensor' is superfluous to requirements .. 


Ramana

**Comment by u99127** [comment]
Not sure why my testing didn't catch this - hopefully this lot satisfies the CI ...
Ramana

File: include/tvm/relay/qnn/attrs.h [hunk 1]
```diff
@@ -177,6 +181,10 @@ struct QnnConv2DAttrs : public tvm::AttrsNode<QnnConv2DAttrs> {
         .describe("The zero point of the input tensor.");
     TVM_ATTR_FIELD(kernel_zero_point)
         .describe("The zero point of the kernel tensor.");
+    TVM_ATTR_FIELD(input_scale)
+      .describe("The quantization scale for the input tensor.");
```
Comment (anijain2305) [inline include/tvm/relay/qnn/attrs.h]: Minor - Can we have same description as Dense here :)
"The input tensor scale."

File: python/tvm/relay/qnn/op/qnn.py [hunk 2]
```diff
@@ -348,7 +367,9 @@ def dense(data,
                        units,
                        input_zero_point,
                        kernel_zero_point,
-                       out_dtype)
+                       out_dtype,
+                       input_scale,
```
Comment (anijain2305) [inline python/tvm/relay/qnn/op/qnn.py]: I think we should keep same order for Python API and C++ API.

**Comment by u99127** [comment]
All updates now done ,please review and merge as appropriate .

**Comment by u99127** [comment]
A gentle ping for a merge.

File: python/tvm/relay/qnn/op/qnn.py [hunk 3]
```diff
@@ -219,6 +221,16 @@ def conv2d(data,
     input_zero_point: int
            The zero point of the data distribution.
 
+    input_scale: float
+           The scale for the input tensor. The scale for the input tensor is
+           stored purely for convenience here. See more commentary below.
+
+    kernel_scale: float
```
Comment (zhiics) [inline python/tvm/relay/qnn/op/qnn.py]: Move these two docstring below `kernel_zero_point` to keep consistent to the order of the parameters.

**Comment by u99127** [comment]
Whoops , now fixed.

**Comment by u99127** [comment]
Ok, it looks like there is some more work to be done here because of other bits that have landed. 
- Fix legalisations.py to consider qnn.conv2d and qnn.dense to filter out the input_scale and kernel_scale before calling nn.conv2d and nn.dense. - Still working on this one.
- Fix qnn.op.dense to have input_scale and kernel_scale- I've done it.

**Comment by anijain2305** [comment]
https://github.com/apache/incubator-tvm/blob/5b9f459d638d1cf6bd820f3bdc58d7e5632d7ed7/python/tvm/relay/qnn/op/legalizations.py#L89-L90

For the first problem, you can just delete the extra scale arguments from the attrs.

**Comment by u99127** [comment]
> https://github.com/apache/incubator-tvm/blob/5b9f459d638d1cf6bd820f3bdc58d7e5632d7ed7/python/tvm/relay/qnn/op/legalizations.py#L89-L90
> 
> For the first problem, you can just delete the extra scale arguments from the attrs.

Ah thanks - I hadn't spotted that last night. 

Now rebased and repushed.

Ramana

**Comment by u99127** [comment]
Now all done, would be nice to merge.

**Comment by zhiics** [comment]
Thanks @u99127 @anijain2305

**Pull request merged**